baselines/pfsp/pfsp_multigpu_cuda.c

```diff
@@ -485,17 +503,86 @@ void pfsp_search(const int inst, const int lb, const int m, const int M, const i
         cudaMemcpy(parents_d, parents, poolSize *sizeof(Node), cudaMemcpyHostToDevice);
 
         // numBounds is the 'size' of the problem
-        evaluate_gpu(jobs, lb, numBounds, nbBlocks, best, lbound1_d, lbound2_d, parents_d, bounds_d);
+        evaluate_gpu(jobs, lb, numBounds, nbBlocks, &best_l, lbound1_d, lbound2_d, parents_d, bounds_d);
 
         cudaMemcpy(bounds, bounds_d, numBounds * sizeof(int), cudaMemcpyDeviceToHost);
 
         /*
-          each task generates and inserts its children nodes to the pool.
+          Each task generates and inserts its children nodes to the pool.
         */
-        generate_children(parents, poolSize, jobs, bounds, &tree, &sol, best, &pool_loc);
+        generate_children(parents, poolSize, jobs, bounds, &tree, &sol, &best_l, pool_loc);
       }
       else {
-        break;
+        // work stealing
+        int tries = 0;
+        bool steal = false;
+        int victims[D];
+        permute(victims,D);
+
+        while (tries < D && steal == false) { // WS0 loop
+          const int victimID = victims[tries];
+
+          if (victimID != gpuID) { // if not me
+            SinglePool_ext* victim;
+            victim = &multiPool[victimID];
+            nSteal ++;
+            int nn = 0;
+            int count = 0;
+            while (nn < 10) { // WS1 loop
+              expected = false;
+              count++;
+              if (atomic_compare_exchange_strong(&(victim->lock), &expected, true)) { // get the lock
+                int size = victim->size;
+                int nodeSize = 0;
+
+                if (size >= 2*m) {
+                  //Node* p = popBackBulkFree(victim, m, M, &nodeSize);
+                  int perc = 2;
```

Comment:
I understand your point. I think treating it like you mentioned is better for the implementation purpose. Either way, since it is still going to depend on the user judgement he could put values greater than 1 (if treating it as values between 0 and 1), or greater than a 100 (if treating it as values between 0 and 100). Maybe I should add as a parameter to be added by the user? Like we do for the instances and set as default 0.5 and restrict values greater than 1 and lower than 0.

Reply:
I propose two options:
- (the lazy one) we hide this to the user for now, and keep 2 (or 0.5). In that case, it could be nice to add a comment in the code saying what is `perc`;
- Move `perc` as a new parameter and update the doc.

Reply:
I think I implemented the second one properly. Let me know.

Reply:
C'est parfait